src/Middleware/DetectLocaleMiddleware.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace TractorCow\Fluent\Middleware;
+
+use SilverStripe\Control\Cookie;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\Middleware\HTTPMiddleware;
+use SilverStripe\Core\Config\Configurable;
+use TractorCow\Fluent\Extension\FluentDirectorExtension;
+use TractorCow\Fluent\Model\Locale;
+use TractorCow\Fluent\State\FluentState;
+use TractorCow\Fluent\State\LocaleDetector;
+
+/**
+ * DetectLocaleMiddleware will detect if a locale has been requested (or is default) and is not the current
+ * locale, and will redirect the user to that locale if needed.
+ */
+class DetectLocaleMiddleware implements HTTPMiddleware
+{
+    use Configurable;
+
+    /**
+     * IDs to persist the locale in cookies / session in the front end, CMS, etc
+     *
+     * @config
+     * @var string[]
+     */
+     #
+    private static $persist_ids = [
+        'frontend' => 'FluentLocale',
+        'cms' => 'FluentLocale_CMS',
+    ];
+
+    /**
+     * The expiry time in days for a locale persistence cookie
+     *
+     * @config
+     * @var int
+     */
+    private static $persist_cookie_expiry = 90;
+
+    /**
+     * Whether cookies have already been set during {@link setPersistLocale()}
+     *
+     * @var bool
+     */
+    protected $cookiesPersisted = false;
+
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        $localeDetector = LocaleDetector::singleton()->setRequest($request);
+
+        $state = FluentState::singleton();
+        $state->setLocale($this->getLocale($request));
+
+        $locale = $state->getLocale();
+        $this->setPersistLocale($request, $locale, $this->getPersistKey());
+
+        return $delegate($request);
+    }
+
+    /**
+     * Get the current locale from routing parameters, persistence, browser locale, etc
+     *
+     * @param  HTTPRequest $request
+     * @return string
+     */
+    public function getLocale(HTTPRequest $request)
+    {
+        $state = FluentState::singleton();
```

Comment:
Do you think it might be useful to separate these middlewares further, so we have one for each: DetectQueryStringLocaleMiddleware, DetectUrlArgumentLocaleMiddleware and DetectPersistedLocaleMiddleware?